src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1456,57 +1437,78 @@ trait Namers extends MethodSynthesis {
       pluginsTypeSig(methSig, typer, ddef, resTpGiven)
     }
 
-    /**
-     * For every default argument, insert a method symbol computing that default
-     */
+    /** Create and enter symbols for default getters. This is done eagerly when the corresponding
+      * method is entered (not later when its type is completed) in order to ensure stability of
+      * scope members (scala/scala#6286 / scala/scala-dev#405).
+      * The symbols are assigned `ErrorType`. The correct type is computed when the method's type is
+      * completed: `methodSig` invokes `addDefaultGetters`, which creates the default getter DefDef
+      * and assigns the correct type to the symbols created here.
+      */
     def enterDefaultGetters(meth: Symbol, ddef: DefDef, vparamss: List[List[ValDef]], tparams: List[TypeDef]): Unit = {
-      val methOwner  = meth.owner
-      val search = DefaultGetterNamerSearch(context, meth, initCompanionModule = false)
+      val owner = meth.owner
+      lazy val ownerNamer: Namer = {
+        val ctx = context.nextEnclosing(c => c.scope.toList.contains(meth)) // TODO use lookup rather than toList.contains
+        assert(ctx != NoContext, meth)
+        newNamer(ctx)
+      }
       var posCounter = 1
-
-      mforeach(vparamss){(vparam) =>
+      mforeach(vparamss) { vparam =>
         // true if the corresponding parameter of the base class has a default argument
         if (vparam.mods.hasDefault) {
           val name = nme.defaultGetterName(meth.name, posCounter)
-
-          search.createAndEnter { owner: Symbol =>
-            methOwner.resetFlag(INTERFACE) // there's a concrete member now
-            val default = owner.newMethodSymbol(name, vparam.pos, paramFlagsToDefaultGetter(meth.flags))
-            default.setPrivateWithin(meth.privateWithin)
-            default.referenced = meth
-            default.setInfo(ErrorType)
-            if (meth.name == nme.apply && meth.hasAllFlags(CASE | SYNTHETIC)) {
-              val att = meth.attachments.get[CaseApplyDefaultGetters].getOrElse({
-                val a = new CaseApplyDefaultGetters()
-                meth.updateAttachment(a)
-                a
-              })
-              att.defaultGetters += default
-            }
-            if (default.owner.isTerm)
-              saveDefaultGetter(meth, default)
-            default
+          owner.resetFlag(INTERFACE) // there's a concrete member now
```

Comment:
Let's move this down a few lines and predicate this on `default.hasFlag(STATIC`)`.

There probably isn't a case now, or even with trait params, when that could actually change behaviour, but it would be slightly clearer to me, especially with an comment.